`lib/syscall-shim`: Support for clone implementation

skuenzer opened the pull request:
This PR introduces support for implementing system calls that 1) require to know the system call return address, like `clone`, 2) support caller TLSes that are different to Unikraft-internal managed TLS, 3) improve the debug printing options, and 4) implement missed features to map system calls statically with `musl`.

GitHub-Depends-On: #564

skuenzer committed:
lib/syscall-shim: Binary system call handler to separate C file

This commit moves the code of the binary system call handler to a
separate C file. This is done so that we can compile this code as
ISR-safe, later.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/syscall-shim: Save & restore ECTX on binary system calls

Compiles the binary system call handler code as ISR-safe. Additionally,
the extended context is saved and restored of the system call caller on
each binary system call. The reason is that extended register sets maybe
used by the system call implementations and the Linux system call
calling convention assumes that caller programs do not have to save the
state themselves.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/syscall-shim: Save & restore TLS pointer on binary system calls

Introduces an option to save & restore the TLS pointer on binary system
calls. Unikraft's internal TLS pointer is loaded before the system call
is handled. This option allows the usage of different TLS pointers
within the system caller program. We name such programs, userland
programs.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/syscall_shim: Compile-time mapping of variadic system calls

Variadic system calls are system calls with a variable number of
arguments. Some of them become optional or are only needed depending on
a given flag or a command number.
This commit introduces the support for statically mapping such
system calls at compile time. For this purpose, the
`uk_syscall_(r_)static()` macros are evaluating the number of given
arguments and will automatically auto-complete missing arguments with
zeros or remove additional arguments that aren't handled by the
particular system call implementation.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/syscall_shim: TLS pointer of caller

If `CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS` is configured, system call
handlers can retrieve the caller's TLS pointer with
`uk_syscall_ultlsp()` if the system call was called by a binary system
call.
Please note that only if the system call was called by a binary system
call and the TLS pointer of the caller is different to the Unikraft-TLS
pointer, this function will return the caller's TLS pointer.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

dragosargint commented on lib/syscall_shim/uk_syscall_binary.c:
```diff
@@ -0,0 +1,100 @@
+	if (likely(ukplat_tlsp_get() == self->uktlsp)) {
+		ukplat_tlsp_set(orig_tlsp);
+	} else {
+		uk_pr_debug("System call updated userland TLS pointer register to %p (before: %p)\n",
```
```suggestion
		uk_pr_debug("System call updated userland TLS pointer register to 0x%lx (before: 0x%lx)\n",
```

skuenzer replied:
A good one! Yes, I'll take it in. 😉

dragosargint reviewed (changes requested):
I have only one suggestion, to change the `uk_pr_debug()` to avoid warnings. Other than that here are some tests that I used [1]
[1] https://github.com/dragosargint/test

skuenzer committed:
lib/syscall_shim: Option to show debug messages on system calls only

This commit introduces two options for printing a debug message 1) when
a system call handler is invoked and 2) when a binary system call is
handled.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukdebug: Always provide `_uk_vprintd()`, `_uk_printd()`

The (internal) symbol declarations of `_uk_vprintd()` and `_uk_printd()`
are always provided. This allows rare cases where libraries have their
own decision logics when and which debug message to compile in.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

adinasm commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -294,6 +326,168 @@ typedef long uk_syscall_arg_t;
+	{								\
+		long ret;						\
+									\
+		ret = rname(UK_ARG_MAPx(x, UK_S_ARG_CAST_LONG, __VA_ARGS__)); \
+		__UK_SYSCALL_RETADDR_ENTRY();				\
```
Shouldn't they be reversed?

skuenzer replied:
Thanks a lot for seeing this. Truely a mistake. The code is updated and changed now.

craciunoiuc reviewed (commented):
I suppose this PR actually contains only the last 2 commits.

Please break up the last commit in 2 (and, as such, shorten the commit message).

Other than this, it looks fine.

skuenzer committed:
lib/syscall_shim: Use build system's AWK rules for generating sources

Uses the build system's ability for generating source files from AWK
programs. This simplifies the library's `Makefile.uk` and the re-
generation is properly triggered on dependency changes.
Because generated header files are stored in a different destination
path, the same change cannot be performed for these files, at the
moment.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/syscall_shim: Caller return address tracking

This commit introduces system call return address tracking, which
is needed by some system call implementations (like `SYS_CLONE`). Such
implementations can query the return address with
`uk_syscall_return_addr()`.

Checkpatch-Ignore: MULTISTATEMENT_MACRO_USE_DO_WHILE
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer commented:
> I suppose this PR actually contains only the last 2 commits.
> 
> Please break up the last commit in 2 (and, as such, shorten the commit message).
> 
> Other than this, it looks fine.

Thanks a lot for chekcing this PR. I updated the commit history. You saw still the draft state. I have the PR final now.

craciunoiuc commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -61,32 +61,87 @@ extern "C" {
+
+#else /* !CONFIG_LIBSYSCALL_SHIM */
+
+/* NOTE: The usage of `uk_syscall_return_addr()` requires the `syscall_shim`
+ *       library to be present. In order to avoid mis-configurations, we
+ *       inject an always failing compile-time assertion to stop compilation.
+ */
```
```suggestion
/**
 * NOTE: The usage of `uk_syscall_return_addr()` requires the `syscall_shim`
 *       library to be present. In order to avoid mis-configurations, we
 *       inject an always failing compile-time assertion to stop compilation.
 */
```

When should we use /** vs /*?

skuenzer replied:
Normally `/**` should be used by Doxygen and to describe interfaces. Since this is a normal comment block, I'd keep the single-asterix version: `/*`. If you think it helps readability, I am applying your suggestion but use `/*`.

craciunoiuc commented on lib/syscall_shim/uk_syscall6.awk:
```diff
@@ -7,13 +7,12 @@ BEGIN {
 	for (i = 1; i < max_args; i++)
-		printf "long arg%d, ",i
-	printf "long arg%d)\n{\n", max_args
+		printf "long arg%d __maybe_unused, ",i
```
```suggestion
		printf "long arg%d __maybe_unused, ", i
```

skuenzer replied:
✅

craciunoiuc commented on lib/syscall_shim/uk_syscall6.awk:
```diff
@@ -7,13 +7,12 @@ BEGIN {
-	printf "long arg%d)\n{\n", max_args
+		printf "long arg%d __maybe_unused, ",i
+	printf "long arg%d __maybe_unused)\n{\n", max_args
+	printf "\tlong ret;\n\n"
```
```suggestion
	print "\tlong ret;\n\n"
```

skuenzer replied:
Updated. Please note that `print` is also issueing a newline while `printf` is not doing this with awk 😉

craciunoiuc commented on lib/syscall_shim/uk_syscall6_r.awk:
```diff
@@ -7,13 +7,12 @@ BEGIN {
 	for (i = 1; i < max_args; i++)
-		printf "long arg%d, ",i
-	printf "long arg%d)\n{\n", max_args
+		printf "long arg%d __maybe_unused, ",i
```
```suggestion
		printf "long arg%d __maybe_unused, ", i
```

skuenzer replied:
Updated ✅

craciunoiuc commented on lib/syscall_shim/gen_syscall_r_static.awk:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		# fill-up missing arguments
+		for (j = i+1; j <= args_nr; j++)
```
```suggestion
		for (j = i + 1; j <= args_nr; j++)
```

skuenzer replied:
Updated.

craciunoiuc commented on lib/syscall_shim/gen_syscall_static.awk:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		# fill-up missing arguments
+		for (j = i+1; j <= args_nr; j++)
```
```suggestion
		for (j = i + 1; j <= args_nr; j++)
```

skuenzer replied:
✅

craciunoiuc commented on lib/syscall_shim/Config.uk:
```diff
@@ -35,15 +35,45 @@ if LIBSYSCALL_SHIM
+			select LIBSYSCALL_SHIM_DEBUG_HANDLER if LIBSYSCALL_SHIM_HANDLER
+			default n
+			help
+				Enable all debug messages of the syscall_shim library
```
```suggestion
				Enable all debug messages of the syscall_shim library.
```

skuenzer replied:
Thanks, I updated this as well :+1:

craciunoiuc reviewed (commented):
Hey @skuenzer,

Left you some style comments. Maybe Codacy should have/has some style checks for awk.

skuenzer committed:
lib/vfscore: Initialize stdout, stdin, stderr as early during init

Instead of initializing the default set of file descriptors with a
constructor function, we initialize them during early init. At this
point of boot time, the heap, scheduling, and TLS are initialized.
Because `vfscore` uses the `dup` system call, we introduce this commit
in preparation for the fact that a TLS may be required to perform a
system call.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer commented:
> Hey @skuenzer,
> 
> Left you some style comments. Maybe Codacy should have/has some style checks for awk.

Hi @craciunoiuc, thanks a lot for finding these style issues. Yeah, I agree it is a pitty that we do not have awk files (and others) covered with checkpatch. I am not sure if `clang-format` could help us with this but I'd assume for awk files we have no luck. 🤷🏻‍♂️ 

Thanks,

Simon

craciunoiuc commented on lib/vfscore/stdio.c:
```diff
@@ -211,15 +212,28 @@ static struct vfscore_file  stdio_file = {
+
+	fd = uk_syscall_r_dup2(0, 1);
+	if (fd != 1) {
+		uk_pr_crit("failed to dup to stdin\n");
```
```suggestion
		uk_pr_crit("failed to dup to stdout\n");
```

Isn't fd==1 stdout or am I misreading the syscall?

skuenzer replied:
I'll update :+1:

craciunoiuc commented on lib/vfscore/stdio.c:
```diff
@@ -211,15 +212,28 @@ static struct vfscore_file  stdio_file = {
 	fd = vfscore_alloc_fd();
-	UK_ASSERT(fd == 0);
+	if (fd != 0) {
+		uk_pr_crit("failed to allocate fd for stdout\n");
```
```suggestion
		uk_pr_crit("failed to allocate fd for stdin\n");
```

Isn't fd == 0 stdin, or am I missing something?

skuenzer replied:
You are right, this was wrong already in the previous code and I did somehow not notice. I am updating this. :+1:

craciunoiuc commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -61,32 +61,107 @@ extern "C" {
+ */
+#define __UK_SYSCALL_RETADDR_CLEAR();					\
+	do { _uk_syscall_return_addr = 0x0; } while (0)
+#endif /* !CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```
```suggestion
#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```

skuenzer replied:
Hum, this is the end of the else case, this why the `!`. We need to document this with the coding style but a lot lof code that we have in the repo follows this principle already. So, I would not change this 😉

craciunoiuc replied:
Ah, okay. I usually see it like "`endif` closes the `ifdef`" and `else` is somewhere in between

okay, you can leave it like this. Hit resolve :)

craciunoiuc commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -61,32 +61,107 @@ extern "C" {
+	do { } while (0)
+#define __UK_SYSCALL_RETADDR_CLEAR();					\
+	do { } while (0)
+#endif /* !CONFIG_LIBSYSCALL_SHIM */
```
```suggestion
#endif /* CONFIG_LIBSYSCALL_SHIM */
```

craciunoiuc commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -61,32 +61,107 @@ extern "C" {
+{
+	return 0x0;
+}
+#endif /* !CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```
```suggestion
#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```

craciunoiuc reviewed (changes requested):
@skuenzer 

Final round of comments from my side, I'll approve after this :)

dragosargint reviewed (approved):
I retested this PR, works well! Thanks, @skuenzer!

Reviewed-by: Dragos Iulian Argint <dragosargint21@gmail.com>

adinasm reviewed (approved):
Looks good!

Reviewed-by: Adina Smeu <adina.smeu@gmail.com>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

skuenzer commented:
> All good here. Thanks!
> 
> Reviewed-by: Cezar Craciunoiu [cezar.craciunoiu@unikraft.io](mailto:cezar.craciunoiu@unikraft.io)

Thanks a lot to you all for checking it!

razvand reviewed (approved):
Thanks, @skuenzer 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
